### PR TITLE
Integrate Hibernate ORM with Hibernate Validator

### LINKED ISSUE #1889: Integrate Hibernate Validator with Hibernate ORM
Normally we expect that a field defined in the Entity class like below

`
@Size(max=100)
@NotEmpty
@Column
private String name;
`

to be generated like that  

`name varchar(100) not null`

but currently hibernate ignores these validation annotations and generates 

`name varchar(255)`

unless we define all constraints with the @Column annotation.

### PR DESCRIPTION
**Pull request by machi1990**
@gsmet this follows up our discussion on Zulip chat on how to go about #1889.
 
Your feedbacks are welcomed as usual :-)

Fixes #1889

### CODE REVIEW

File: extensions/hibernate-validator/runtime/src/main/java/io/quarkus/hibernate/validator/runtime/HibernateValidatorTemplate.java
Comment (gsmet) [inline extensions/hibernate-validator/runtime/src/main/java/io/quarkus/hibernate/validator/runtime/HibernateValidatorTemplate.java]: I don't understand why you define 2 different validators?

**Review by gsmet — COMMENTED** [review]
@machi1990 thanks!

I added an important question inline. Could you explain your rationale?

File: extensions/hibernate-orm/deployment/src/test/java/io/quarkus/hibernate/orm/validation/JPAValidationTestCase.java
Comment (machi1990) [reply extensions/hibernate-orm/deployment/src/test/java/io/quarkus/hibernate/orm/validation/JPAValidationTestCase.java]: Sure, I'll map the thrown exception into something appropriate.

File: extensions/hibernate-orm/deployment/pom.xml
Comment (machi1990) [reply extensions/hibernate-orm/deployment/pom.xml]: Done.

File: extensions/hibernate-validator/runtime/src/main/java/io/quarkus/hibernate/validator/runtime/HibernateValidatorTemplate.java
Comment (machi1990) [reply extensions/hibernate-validator/runtime/src/main/java/io/quarkus/hibernate/validator/runtime/HibernateValidatorTemplate.java]: It is because I was getting this 
```
java.lang.IllegalStateException: Defining a Validator-specific traversable resolver is not supported by the predefined scope ValidatorFactory.
```

when trying to use `PredefinedScopeHibernateValidatorFactory`. 

Any ideas?

Comment (gsmet) [reply extensions/hibernate-validator/runtime/src/main/java/io/quarkus/hibernate/validator/runtime/HibernateValidatorTemplate.java]: Hmmm, right. So that's something we need to fix in HV. I was a bit too conservative on what would be accepted.

TBH, I didn't think the traversable resolver would be set for each JPA validation call.

Could you change your code to use the initial `ValidatorFactory` (well, it will mostly be a matter of reverting changes)? I'll push a PR to HV so that you can test if it works OK for you.

Comment (gsmet) [reply extensions/hibernate-validator/runtime/src/main/java/io/quarkus/hibernate/validator/runtime/HibernateValidatorTemplate.java]: @machi1990 can you check with this branch of Hibernate Validator: https://github.com/hibernate/hibernate-validator/pull/1039 ?

It should solve the issue and you should be able to have only one VF.

I'll wait for your feedback before merging and releasing a new alpha.

Comment (machi1990) [reply extensions/hibernate-validator/runtime/src/main/java/io/quarkus/hibernate/validator/runtime/HibernateValidatorTemplate.java]: > @machi1990 can you check with this branch of Hibernate Validator: [hibernate/hibernate-validator#1039](https://github.com/hibernate/hibernate-validator/pull/1039) ?
> 
> It should solve the issue and you should be able to have only one VF.
> 
> I'll wait for your feedback before merging and releasing a new alpha.

@gsmet thanks for the quick response. It solves the problem.

File: extensions/hibernate-orm/deployment/src/test/java/io/quarkus/hibernate/orm/validation/JPAValidationTestCase.java
Comment (machi1990) [reply extensions/hibernate-orm/deployment/src/test/java/io/quarkus/hibernate/orm/validation/JPAValidationTestCase.java]: Done

File: extensions/hibernate-validator/runtime/src/main/java/io/quarkus/hibernate/validator/runtime/HibernateValidatorTemplate.java
Comment (machi1990) [reply extensions/hibernate-validator/runtime/src/main/java/io/quarkus/hibernate/validator/runtime/HibernateValidatorTemplate.java]: @gsmet Hi, any news on the HV release?

Comment (gsmet) [reply extensions/hibernate-validator/runtime/src/main/java/io/quarkus/hibernate/validator/runtime/HibernateValidatorTemplate.java]: Sorry, still in the pipeline. I'm preparing a talk and didn't find the time to do it.

I think I should be able to push the release button this afternoon.

Comment (machi1990) [reply extensions/hibernate-validator/runtime/src/main/java/io/quarkus/hibernate/validator/runtime/HibernateValidatorTemplate.java]: Cool :-)

Comment (gsmet) [reply extensions/hibernate-validator/runtime/src/main/java/io/quarkus/hibernate/validator/runtime/HibernateValidatorTemplate.java]: OK, release done. Now, we have to wait for it to be synced to Central. Hopefully, it will be synced today and I'll push an upgrade PR to fix another issue and prepare for this one.

**Comment by machi1990** [comment]
@gsmet finally, we are ready for a full review :-)

**Comment by gsmet** [comment]
@machi1990 just a status: I worked on this one yesterday, trying to get the bean manager working. It's a bit more complicated than expected to get things working.

I hope I'll get something working before the release.

**Comment by machi1990** [comment]
> @machi1990 just a status: I worked on this one yesterday, trying to get the bean manager working. It's a bit more complicated than expected to get things working.
> 
> I hope I'll get something working before the release.

Thanks for the update. Hope we get it working before the release.And then I'll be interested in finding out how you did it.

**Review by gsmet — APPROVED** [review]
I won't have the time to finish the bean manager part before the release and this is already a big improvement so let's merge it.

**Comment by gsmet** [comment]
Thanks @machi1990 ! I'll ping you when I create the bean manager PR.

**Pull request merged**
